### PR TITLE
Add ED test to Self-chat

### PR DESCRIPTION
**Pull request by dianaglzrico**
**Patch description**
Self-chat was breaking when  seed utterances but not context were provided like [in here](https://github.com/facebookresearch/ParlAI/issues/3064), as `self_observe()` was being called without an `observe()`. An ED test was added that resembles this case to verify the fix.

### CODE REVIEW

File: parlai/tasks/empathetic_dialogues/worlds.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class SelfChatWorld(SelfChatBaseWorld):
+    def __init__(self, opt, agents, shared=None):
+        super().__init__(opt, agents, shared)
+
+    def get_contexts(self):
+        return None
+
+    def share(self):
```
Comment (stephenroller) [inline parlai/tasks/empathetic_dialogues/worlds.py:17]: This is a no-op

File: parlai/tasks/empathetic_dialogues/worlds.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from parlai.tasks.self_chat.worlds import SelfChatWorld as SelfChatBaseWorld
+
+
+class SelfChatWorld(SelfChatBaseWorld):
+    def __init__(self, opt, agents, shared=None):
```
Comment (stephenroller) [inline parlai/tasks/empathetic_dialogues/worlds.py:11]: So is this

File: tests/test_self_chat.py [hunk 3]
```diff
@@ -29,3 +29,16 @@
...
+        opt = pp.parse_args(
+            [
+                '-mf',
+                'zoo:blender/blender_3B/model',
+                '-t',
+                'empathetic_dialogues',
+                '--seed-messages-from-task',
+            ]
+        )
+        self_chat.self_chat(opt)
```
Comment (stephenroller) [inline tests/test_self_chat.py:44]: Can we use SelfChat.main(task='empathetic_dialogues', ...)

File: tests/test_self_chat.py [hunk 4]
```diff
@@ -29,3 +29,16 @@
             ]
         )
         self_chat.self_chat(opt)
+
+    def test_ed(self):
+        pp = self_chat.setup_args()
+        opt = pp.parse_args(
+            [
+                '-mf',
+                'zoo:blender/blender_3B/model',
```
Comment (stephenroller) [inline tests/test_self_chat.py:38]: This is a very heavy model for a small test. Can we use the tutorial generator?

File: tests/test_self_chat.py [hunk 5]
```diff
@@ -29,3 +29,16 @@
...
+                '-t',
+                'empathetic_dialogues',
+                '--seed-messages-from-task',
```
Comment (stephenroller) [inline tests/test_self_chat.py:41]: Does this not need a true argument?

**Review by stephenroller — COMMENTED** [review]
A few questions

File: parlai/tasks/empathetic_dialogues/worlds.py (hunk 1 above)
Comment (dianaglzrico) [reply parlai/tasks/empathetic_dialogues/worlds.py:17]: fixed

File: parlai/tasks/empathetic_dialogues/worlds.py (hunk 2 above)
Comment (dianaglzrico) [reply parlai/tasks/empathetic_dialogues/worlds.py:11]: fixed

File: tests/test_self_chat.py (hunk 3 above)
Comment (dianaglzrico) [reply tests/test_self_chat.py:44]: Why or what would be the difference to use this?

File: tests/test_self_chat.py (hunk 4 above)
Comment (dianaglzrico) [reply tests/test_self_chat.py:38]: Yeah you're right about that, will change it.

File: tests/test_self_chat.py (hunk 5 above)
Comment (dianaglzrico) [reply tests/test_self_chat.py:41]: no this is good, it's a `store_true` flag

File: tests/test_self_chat.py (hunk 4 above)
Comment (dianaglzrico) [reply tests/test_self_chat.py:38]: Yeah you're right, I'll change it

File: tests/test_self_chat.py (hunk 3 above)
Comment (stephenroller) [reply tests/test_self_chat.py:44]: Cleaner and newer syntax. Functionally equivalent.

File: tests/test_self_chat.py (hunk 5 above)
Comment (stephenroller) [reply tests/test_self_chat.py:41]: Can we also change that to not be a store_true then? Store_true is impossible to override once it's been set in an opt file

Comment (dianaglzrico) [reply tests/test_self_chat.py:41]: Yeah, that's true, I'll change it.

Comment (dianaglzrico) [reply tests/test_self_chat.py]: Addressed this [here](https://github.com/facebookresearch/ParlAI/pull/3113).

**Review by emilydinan — COMMENTED** [review]
i don't really understand the difference between openers and context.. can you explain @dianaglzrico ? should we change teh default in the base world?

**Comment by dianaglzrico** [comment]
> eally understand the difference between openers and context.. can y

As I understand in self chat, context is used to provide info depending on the task (personas, topics, etc) and openers which can be taken from the task too but are seeded messages used to open the conversation. The issue that I linked in the description reported an error when you provide openers and no context at all.

**Review by stephenroller — APPROVED** [review]
Thank you

**Pull request merged**